internal/graph/iterators.go

```diff
@@ -0,0 +1,75 @@
+package graph
+
+import (
+	"context"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+// cachedUserTuple is a simple struct to hold the minimal data needed
+// to store in a cached iterator.
+type cachedUserTuple struct {
+	user      string
+	condition *openfgav1.RelationshipCondition
+	timestamp *timestamppb.Timestamp
+}
+
+// cachedUserTupleIterator is a wrapper around a cached iterator
+// for a given object/relation.
+type cachedUserTupleIterator struct {
+	object   string
+	relation string
+	iter     storage.Iterator[cachedUserTuple]
+}
```

Comment:
I'm curious on why we are changing the code one level above of where we could be making the change (the storage interface itself). We could be reducing network transfers for everyone, regardless of whether you have caching turned on.

Reply:
I'm not following, what are you suggesting?

Reply:
Change `storage.ReadUsersetTuples` so that it returns `UserTupleIterator`

Reply:
> We could be reducing network transfers for everyone, regardless of whether they have caching turned on.
I'm still not following, how would "reducing network transfers" work without caching turned on in this case?

Reply:
We change the storage layer itself rather than the caching layer above it. We make the storage layer return an optimized iterator :)